pkg/agent/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,439 @@
+		time.Sleep(uploadToFileServerRetryDelay)
+	}
+	return nil
+}
+
+func (c *SupportBundleController) uploadToFileServer(bundleName string, url string, serverAuth *cpv1b2.BundleServerAuthConfiguration, tarGzFile io.Reader) error {
+	strArr := strings.Split(url, "/")
+	address := strArr[0]
+	targetPath := url[len(address):]
+	path := path.Join(targetPath, c.nodeName+"_"+bundleName+".tar.gz")
+	cfg := &ssh.ClientConfig{
+		User: serverAuth.BasicAuthentication.Username,
+		Auth: []ssh.AuthMethod{ssh.Password(serverAuth.BasicAuthentication.Password)},
+		// #nosec G106: skip host key check here and users can specify their own checks if needed
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         time.Second,
+	}
+	uploader, err := getUploader(sftpProtocol)
+	if err != nil {
+		return err
+	}
+	return uploader.upload(address, path, cfg, tarGzFile)
+}
+
+func getUploaderByProtocol(protocol ProtocolType) (uploader, error) {
+	if protocol == sftpProtocol {
+		if sftpUploaderInstance == nil {
+			sftpUploaderInstance = &sftpUploader{}
+		}
+		return sftpUploaderInstance, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+type uploader interface {
+	upload(addr string, path string, config *ssh.ClientConfig, tarGzFile io.Reader) error
+}
+
+type sftpUploader struct {
+}
+
+func (uploader *sftpUploader) upload(address string, path string, config *ssh.ClientConfig, tarGzFile io.Reader) error {
+	conn, err := ssh.Dial("tcp", address, config)
+	if err != nil {
+		return fmt.Errorf("error when connecting to fs server: %w", err)
+	}
+	sftpClient, err := sftp.NewClient(conn)
+	if err != nil {
+		return fmt.Errorf("error when setting up sftp client: %w", err)
+	}
+	defer func() {
+		if err := sftpClient.Close(); err != nil {
+			klog.Errorf("Error when closing sftp client: %v", err)
+		}
+	}()
+	targetFile, err := sftpClient.Create(path)
+	if err != nil {
+		return fmt.Errorf("error when creating target file on remote: %v", err)
+	}
+	defer func() {
+		if err := targetFile.Close(); err != nil {
+			klog.Errorf("error when closing target file on remote: %v", err)
+		}
+	}()
+	if written, err := io.Copy(targetFile, tarGzFile); err != nil {
+		return fmt.Errorf("error when copying target file: %v, written: %d", err, written)
+	}
+	klog.InfoS("Successfully upload file to path", "filePath", path)
+	return nil
+}
+
+func (c *SupportBundleController) updateSupportBundleCollectionCondition(condition v1alpha1.SupportBundleCollectionConditionType) {
+	c.conditionLock.Lock()
+	defer c.conditionLock.Unlock()
+	c.collectionCondition = condition
+}
+
+func (c *SupportBundleController) updateSupportBundleCollectionStatus(key string, complete bool, genErr error) error {
+	antreaClient, err := c.antreaClientGetter.GetAntreaClient()
+	if err != nil {
+		return fmt.Errorf("failed to get antrea client: %w", err)
+	}
+	var nodeType string
+	if c.nodeType == config.K8sNode {
```

Comment:
Shall we use "controlplane.SupportBundleCollectionNodeType" as the value of SupportBundleController.nodeType ?